text/0000-abstract-prerender.md

```diff
@@ -0,0 +1,84 @@
+- Start Date: 2021-05-29
+- RFC PR: 
+- Svelte Issue: 
+
+# Abstract pre-render
+
+## Summary
+
+Adding support for prerendered HTML files which look for their page store data at client-side runtime (what I'd call "abstract"), instead of using the snapshot of such data as it was at build time (what I'd call "concrete").
+
+## Motivation
+
+This will allow us to pre-render every page, even the ones that do something with the request-specific page store data in `onMount`. It is possible to do this today with no changes to SvelteKit itself, if you have a web server that can read and transform (interpolating values) into the html artifacts for each request. But doing that is messy.
+
+There's a lot more detail about motivation in the initial comment of https://github.com/sveltejs/kit/issues/1533.
+
+## Detailed design
+
+### Technical Background
+
+Right now, a pre-rendered page has code like this to hydrate (this is for a parameterized route):
+
+    page: {
+      host: location.host, // TODO this is redundant
+      path: "/my/page/123",
+      query: new URLSearchParams("foo=bar"),
+      params: {"id": "123"}
+    }
+
+Aside from the "host", those are all the concrete, specific values as they were at build time, baked into the page as literals. But one might want to use the prerendered HTML artifact in an abstract way, for variable and wildcard request parameters, in which case you'd want something more like this:
+
+    page: {
```

Comment:
You don't have to use the page store. You can just use `location` on the client-side. I don't think we'd want the page store on the client to be different from the page store on the server

Reply:
Sure, for query strings you could circumvent the page store and look at location.search, but if you're using ssr and client routing, doesn't that make a weird situation where, on initial mount you have to use location but on subsequent pages you can use the page store? Seems more consistent if they can just always use the page store.

Also, for path params, just using location wouldn't be an option because the client doesn't have the pattern or param names.

Reply:
The query string and host aren't available at prerender time. You have to get them from the `location`. It's better to make this explicit

Reply:
That's fine, it's just that pesky path params then. I wonder if we could monkey patch a "params" property onto the `window.location` object. I mean of course we could, I wonder if you think we should? That would allow "location" to give a complete, dynamic view of everything the page store would normally contain. It's kind of abusing the Location builtin, and people who use Typescript might have trouble accessing a property that's not supposed to be on Location, but aside from that seems pretty elegant to me.

Reply:
Though, I guess this wouldn't work for people using client routing and static ssr. They'd have to look at `location.params` on the first page, and `$page.store.params` on subsequent pages.

Reply:
Regarding https://github.com/sveltejs/rfcs/pull/36, maybe that's the answer. If you need to know the path params from `onMount` in a static pre-rendered page (at client runtime), instead of page store or location, maybe it's `routerInstance.current().params` or similar. Just want to make sure this all works even for people not using client-side transitions, e.g. `router: false` in the Svelte config.